tests/phpunit/tests/kses.php

```diff
@@ -1430,6 +1419,79 @@ public static function data_possible_custom_data_attributes_and_transformed_name
 		);
 	}
 
+	/**
+	 * Ensures that only allowable custom data attributes are retained.
+	 *
+	 * @ticket 33121
+	 *
+	 * @dataProvider data_data_attributes_and_whether_they_are_allowed
+	 *
+	 * @param string $attribute_name Custom data attribute, e.g. "data-wp-bind--enabled".
+	 * @param bool   $is_allowed     Whether the given attribute should be allowed.
+	 */
+	public function test_wp_kses_attr_data_attribute_is_allowed( $attribute_name, $is_allowed ) {
+		$element = "<div {$attribute_name}>Pens and pencils.</div>";
+
+		$processor = new WP_HTML_Tag_Processor( $element );
+		$processor->next_tag();
+
+		$this->assertTrue(
+			$processor->get_attribute( $attribute_name ),
+			"Failed to find expected attribute '{$attribute_name}' before filtering: check test."
+		);
+
+		$processor = new WP_HTML_Tag_Processor( wp_kses_post( $element ) );
+		$this->assertTrue(
+			$processor->next_tag(),
+			'Failed to find containing tag after filtering: check test.'
+		);
+
+		if ( $is_allowed ) {
+			$this->assertTrue(
+				$processor->get_attribute( $attribute_name ),
+				"Allowed custom data attribute '{$attribute_name}' should not have been removed."
+			);
+		} else {
+			$this->assertNull(
+				$processor->get_attribute( $attribute_name ),
+				"Should have removed un-allowed custom data attribute '{$attribute_name}'."
+			);
+		}
```

Comment:
I know that @westonruter and I have discussed moving more tests to the HTML API since that can help us pinpoint the things we care about and avoid test fragility.

For example, is it a problem if we get `<div >` instead of `<div>`? No, that doesn't indicate that the test failed, only that something neutral occurred.

We have some tests using `DOMDocument` for this purpose `assertEquivalentMarkup`, but `DOMDocument` is so flawed it's hard to recommend it for testing. The HTML API so far is tested for spec-compliance, so ideally if there were a bug elsewhere it would cause failure first in the HTML API tests.

Not going to take stand on this one, but this is essentially testing `$this->assertObjectHasProperty()` vs `$this->assertSame()` and I wrote it this way to ensure the test doesn't accidentally fail when the code-under-test is still working. (This has been a great frustrating when trying to improve Core's HTML-handling code, as most changes that fix bugs involve updating tens or hundreds of unit tests that have over-specified the output).

Reply:
> We have some tests using `DOMDocument` for this purpose `assertEquivalentMarkup`, but `DOMDocument` is so flawed it's hard to recommend it for testing. The HTML API so far is tested for spec-compliance, so ideally if there were a bug elsewhere it would cause failure first in the HTML API tests.
> 
> Not going to take stand on this one, but this is essentially testing `$this->assertObjectHasProperty()` vs `$this->assertSame()` and I wrote it this way to ensure the test doesn't accidentally fail when the code-under-test is still working. (This has been a great frustrating when trying to improve Core's HTML-handling code, as most changes that fix bugs involve updating tens or hundreds of unit tests that have over-specified the output).

It's not a hill I'm going to die on either.  :)

I'm happy to keep this unchanged.